buildindex.1
.\" buildindex.1 -- 
.\" Copyright 1994, 1995 Rik Faith (faith@cs.unc.edu)
.\"
.\" This program is free software; you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by the
.\" Free Software Foundation; either version 1, or (at your option) any
.\" later version.
.\"
.\" This program is distributed in the hope that it will be useful, but
.\" WITHOUT ANY WARRANTY; without even the implied warranty of
.\" MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
.\" General Public License for more details.
.\"
.\" You should have received a copy of the GNU General Public License along
.\" with this program; if not, write to the Free Software Foundation, Inc.,
.\" 675 Mass Ave, Cambridge, MA 02139, USA.
.\"
.\" NOTE: This program is not distributed with any data.  This program
.\" operates on many different data types.  You are responsible for
.\" obtaining any data upon which this program operates.
.\" "
.TH BUILDINDEX 1 "23 Aug 1995" "" ""
.SH NAME
buildindex \- build an index file for the dict program
.SH SYNOPSIS
.BI "buildindex [\-Djdlr] " input_file
.SH DESCRIPTION
.B buildindex
takes as input dictionaries in various formats and output an index which is
compatibale with
.BR dict (1).
The index file has lines with the headword and offset separated by a tab
character.  The offset is in decimal.  The headword must not contain any
tab characters.  Each line ends with a newline.
.SH OPTIONS
.TP
.B \-j
Assumes that the input file has the same format as the Jargon File.  That
is, all headwords are bracketed by ':' characters, and each entry ends on
the line prior to the next line that starts with a ':' character.
.sp
The Jargon File is available from prep.ai.mit.edu:/pub/gnu/jarg310.txt.gz
.TP
.B \-d
Assumes that the input file has the same format as The Devil's Dictionary
(by Amrose Bierce), as distributed by the Internet Wiretap.  That is, all
headwords are in uppercase and end with a '.' or ',' character.  An entry
ends with a blank line prior to the beginning of the next entry.
.sp
The Devil's Dictionary is available from
wiretap.spies.com:/Library/Classic/devils.txt
.TP
.B \-f
Assumes that the input file has the same format as the Free On-line
Dictionary of Computing (FOLDOC) from Denis Howe.  That is, all headwords
start in column one and have a newline before them.  All entries end with
the start of the next headword.
.sp
FOLDOC is available from wombat.doc.ic.ac.uk:/pub/Dictionary
.TP
.B \-l
Assumes that headwords start in the first column and continue until some
non-space punctuation is encountered.  Entries end with a blank line.
.TP
.B \-r
Reformats a file with the format described for
.BR \-l ,
inserting newlines in between entries.  This is necessary for
.BR dict (1),
since
.BR dict (1)
can't differentiate between the original format and other, similar formats.
Both
.B \-l
and
.B \-r
are for a latin dictionary that was posted to a newsgroup a long time ago.
It does not appear to be available on the net at this time.
.TP
.B \-D
Turns on debugging.  This options probably doesn't do much.
.SH "SEE ALSO"
.BR dict (1),
.BR compressdict (1)
.SH AUTHOR
The
.BR buildindex (1)
program was written by Rickard E. Faith (faith@cs.unc.edu) and is
distributable under the terms of the GNU General Public License.
.sp
The end-user is responsible for obtaining all data upon which this program
operates.  Please do not mail the author with requests for data.


NEWS
Version 1.9.13, by Aleksey Cheusov, 24 Mar 2004 11:32:53 +0200

  FIXED: if database has no 00-database-alphabet headword,
      "global" alphabet is built incorrectly. As a result
      dictd may work (FreeBSD does) incorrectly with LEV strategy.
      Affected versions: 1.9.12

  ADDED: plugin using libdbi library for implementing DICT database using
    SQL server. It is possible to specify custom search strategies
    by specifying SQL query.

    The following option are currently supported:

      driverdir       - path to DBI drivers
      drivername      - DBI driver name, such as pgsql, mysql, sqlite etc.
      option_host     - host to connect to
      option_port     - port to connect to
      option_dbname   - database name
      option_username - user name for authorizing to SQL server
      option_password - user password for authorizing to SQL server
      query_define    -        SQL query for implementing DEFINE command
      query_<strategy> - SQL query for implementing
           'strategy' (MATCH command).
           Note that 'strategy' is not limited to the "standard" ones.

    See the file example_plugin_dbi.conf for sample of how
      this plugin can be configured.

    SQL queries should return 1-column result by using SELECT command.
      %q sequence inside SQL query is expanded to user's query.
      %% is expanded to single % sign.

    This plugin gives a simplest way to implement editable
      dictionaries with a help
      of WEB interface, but with a read-only access to it
      using DICT protocol.

    NOTE! Unless libdbi library is not found on your system
    or you configure dictd with --disable-plugin option,
    'dictd' is linked with libdbi to resolv DBI driver's symbols.
    This should be fixed.

    NOTE! Connection to SQL database is made
    for every DEFINE/MATCH command.
    This slows down plugin and should be fixed.

Version 1.9.12, by Aleksey Cheusov, 10 Mar 2004 18:45:15 +0200

  Both dictd and Judy-plugin correctly implement
    the 'lev' strategy for utf-8 databases.
    For this purpose an additional headword 00-database-alphabet is
    created by dictfmt.
    So, if you often use 'lev' strategy with utf-8 dictionaries,
    rebuild them using dictfmt-1.9.12

  'nl_langinfo' function is used to detect
      whether dictd/dictfmt uses 8bit, utf-8, or C locale.

  dictd:
    - fixes

       SHOW SERVER command cuts off long database names

       crash on Solaris (and FreeBSD5 ?) due to
         toupper('255') returning 376 if utf-8 locale is set.

       LC_CTYPE and LC_COLLATE are set according to --locale.
         As a result date and time are logged using the C locale

    - MATCH|DEFINE *|! ignore invisible dictionaries.
         Now there is no reason to separate visible and invisible dictionaries
         by database_exit command.
         This simplifies the configuration file.

    - dictd --test|--test-file notifies about invalid database name

  dictl:

    - checks whether recode/iconv/konwert/locale commands are available

    - analyses PAGER variable to set a pager.

    - incorrect charset determination has been fixed

  dictfmt:

     - new option --break-headwords for use with --headword-separator, so
       each headword is written on its own line of the .dict file.

     - correctly wraps utf-8 input

     - new option --columns to set a number of columns for
      wrapping text. See dictfmt.1. By default this value is 72.
      All formats are affected and use the same default.

     - 00-database-XXX headwords are not copied to .dict file.
         The only exception is 00-database-short which should be copied
         to .dict file for older 'dictd'

  maa:
      new functions and #defines:
         'str_pool_init_position', 'str_pool_next_position',
         'str_pool_get_position', 'str_pool_readonly', 'str_pool_iterate',
         'str_pool_iterate_arg', 'STR_ITERATE' and 'STR_ITERATE_END'
         to iterate over strings in the string pool.

      fixed a bug in flags.c/flg_set function

  dictdplugin_judy:
      
      fixed: if a requested headword has more than one definition,
         DEFINE command always returns the first one

Version 1.9.11, by Aleksey Cheusov, 8 Dec 2003 20:55:31 +0200

  NEWS file added

  dictd:
    - dictd is compiled successfully on Solaris, MacOS-X(cc).
      A lot of warning and error messages appeared
      on 64-bit Opteron and Alpha were fixed.

    - fixed: 'suffix' strategy doesn't ignore characters other than
       alphanumeric and spaces. Affected versions: 1.9.1-1.9.10
    - A multi-lined short name of the database is truncated before sending
      to the client

    - patch by Michael Bunk <bunk at imn.htwk-leipzig.de} fixing:
       - a typo giving the incomprehensible failure message.
         Affected versions: 1.7.1-1.9.10

       - a typo resulting in 'dict --server-info' showing
           same compressed/uncompressed database sizes
         Affected versions: 1.9.1-1.9.10

       - broken AUTH command.
         Affected versions: 1.9.4-1.9.10

    - added: keyword "disable_strategy for disabling a specified strategy
             for a particular database.
             Strategies disabled by 'disable_strategy' keyword
             are not passed to plugins.

    - info/name keywords in 'database_virtual' section can handle
      entry name beginning with @ symbol in a similar way
      as 'database' and 'database_plugin' do.

    - dictd --pp <prog> sets a preprocessor for configuration file.

      If you run DICT server having a lot of dictionaries, you can preprocess
      configuration file using m4 (or cpp). This significantly reduces
      the size of configuration file.

      See dictd.8 and example_complex.conf.

  plugins:
    - the default directory for databases
      is passed to plugins (DICT_PLUGIN_INITDATA_DEFDBDIR)

    - plugin support is ported to systems having dlXXX API in libc (FreeBSD).


buildindex.c
/* buildindex.c -- Buildindex
 * Created: Sat Nov 26 23:38:07 1994 by faith@cs.unc.edu
 * Revised: Fri Dec  2 17:42:05 1994 by faith@cs.unc.edu
 * Copyright 1994 Rickard E. Faith (faith@cs.unc.edu)
 *
 * This program is free software; you can redistribute it and/or modify it
 * under the terms of the GNU General Public License as published by the
 * Free Software Foundation; either version 1, or (at your option) any
 * later version.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 * NOTE: This program is not distributed with any data.  This program
 * operates on many different data types.  You are responsible for
 * obtaining any data upon which this program operates.
 *
 */

#include <stdio.h>
#include <string.h>
#include <ctype.h>
#ifdef __linux__
#include <getopt.h>
#else
extern int optind;
#endif

#define buffer_size 1024

static int Debug;
static int jargon;
static int devils;
static int foldoc;
static int latin;
static int reformat;

static void add( FILE *str, char *word, unsigned long offset )
{
   fprintf( str, "%s\t%lu\n", word, offset );
}

int main( int argc, char **argv )
{
   FILE          *infile;
   FILE          *outfile;
   char          buffer[buffer_size];
   int           c;
   int           len;
   unsigned long lines     = 0;
   unsigned long headwords = 0;
   int           prev      = 0;

   while ((c = getopt( argc, argv, "Djdflr" )) != EOF)
	 switch (c) {
	 case 'D': ++Debug;    break;
	 case 'j': ++jargon;   break;
	 case 'd': ++devils;   break;
	 case 'f': ++foldoc;   break;
	 case 'l': ++latin;    break;
	 case 'r': ++reformat; break;
	 }
   
   if (argc - optind != 1
       || (!jargon && !devils && !foldoc && !latin && !reformat)) {
      fprintf( stderr, "usage: buildindex [-Djdflr] input_file\n" );
      return 1;
   }

   if (!(infile = fopen( argv[2], "r" ))) {
      fprintf( stderr, "Cannot open \"%s\" for input\n", argv[2] );
      return 1;
   }

   if (reformat) {
      int first = 1;
      
      while (fgets( buffer, buffer_size, infile )) {
	 if (first) first = 0;
	 else if (buffer[0] != ' ') printf( "\n" );
	 printf( "%s", buffer );
      }

      fclose( infile );
      return 0;
   }

   sprintf( buffer, "sort -d -f > %s.index", argv[optind] );

   if (!(outfile = popen( buffer, "w" ))) {
      fprintf( stderr, "Cannot open pipe for: \"%s\"\n", buffer );
      return 1;
   }

   while (fgets( buffer, buffer_size, infile )) {
      ++lines;
      len = strlen( buffer );
      
      if (!(lines % 1000)) {
	 fprintf( stderr, "\r%10lu lines, %lu headwords", lines, headwords );
      }
      
      if (jargon && buffer[0] == ':') {
	 char *pt = strchr( buffer + 1, ':' );

	 if (pt) {
	    *++pt = '\0';
	    add( outfile, buffer, ftell( infile ) - len );
	    ++headwords;
	 }
      }

      if (devils && prev == '\n' && isupper( buffer[0] )) {
	 char *pt = strchr( buffer, ',' );

	 if (!pt) pt = strchr( buffer, '.' );

	 if (pt) {
	    *pt = '\0';
	    add( outfile, buffer, ftell( infile ) - len );
	    ++headwords;
	 }
      }

      if (foldoc && prev == '\n' && buffer[0] != ' ' && buffer[0] != '\t') {
	 char *pt = strchr( buffer, '\n' );

	 if (pt) {
	    *pt = '\0';
	    add( outfile, buffer, ftell( infile ) - len );
	    ++headwords;
	 }
      }

      if (latin && buffer[0] != '\n' && buffer[0] != ' ') {
	 char *pt = buffer;

	 while (pt && (*pt == ' ' || isalpha( *pt ))) ++pt;
	 if (pt && pt[-1] == ' ') --pt;
	 while (pt && *pt == ' ') --pt;
	 if (pt && isalpha( *pt )) ++pt;

	 if (pt) {
	    *pt = '\0';
	    if (strlen( buffer )) {
	       add( outfile, buffer, ftell( infile ) - len );
	       ++headwords;
	    }
	 }
      }

      prev = buffer[0];
   }

   pclose( outfile );
   fclose( infile );

   fprintf( stderr, "\r%10lu lines, %lu headwords total\n", lines, headwords );

   return 0;
}
